File: CODING/8. Trees/Check_sum_tree.c
/*
Program to check the Binary Tree is Sum Tree or not....

Time Complexity : O(n)
Space Complexity : O(n)
*/

#include<stdio.h>
#include<stdlib.h>
#include<math.h>

struct node
{
	int data;
	struct node *left,*right;
};

struct node *newNode(int val)
{
	struct node *nn = (struct node*)malloc(sizeof(struct node));
	
	nn->data = val;
	
	nn->left = nn->right = NULL;
	
	return nn;
}

int checkSumTree(struct node *root)
{
	if(!root)
		return 0;
		
	if(!root->left && !root->right)
		return root->data;
		
	int leftSum = checkSumTree(root->left);
	
	int rightSum = checkSumTree(root->right);
	
	if(!leftSum || !rightSum)
		return 0;
	
	if(root->data == leftSum + rightSum)
		return 2 * root->data;
	
	return 0;
}

void main()
{
	struct node *root = newNode(50);
	root->left = newNode(15);
	root->right = newNode(10);
	root->left->left = newNode(6);
	//root->left->left->left = newNode(5);
	root->left->right = newNode(9);
	root->right->left = newNode(4);
	root->right->right = newNode(6);
	
	printf(checkSumTree(root) ? "Yes" : "No");
}

File: interview/src/tree/CheckSumTree.java
/**
 * https://www.geeksforgeeks.org/check-if-a-given-binary-tree-is-sumtree/
 *
 * Time Complexity : O(N)
 * Space Complexity : O(N)
 * */
package tree;

public class CheckSumTree {
        
        private int checkSumTree(TreeNode root) {
                // base case
                if(root == null) {
                        return 0;
                }
                // if leaf
                if(root.left == null && root.right == null) {
                        return root.data;
                }
                // capture sum of left and right tree
                int leftTreeSum = checkSumTree(root.left);
                int rightTreeSum = checkSumTree(root.right);
                // if current subtree is sum tree
                if(leftTreeSum != -1 && rightTreeSum != -1 && (leftTreeSum + rightTreeSum == root.data)) {
                        return 2 * root.data;
                }
                return -1;
        }
        
        public static void main(String[] args) {
                TreeNode root = new TreeNode(26);
                root.left = new TreeNode(10);
                root.right = new TreeNode(3);
                root.left.left = new TreeNode(4);
                root.left.right = new TreeNode(6);
                root.right.right=new TreeNode(3);
                System.out.println(new CheckSumTree().checkSumTree(root) != -1 ? "YES" : " NO");
        }
}
